### PR TITLE
Get local resolver from stack.yaml file

### LINKED ISSUE #125: Support for hash value Stack resolver
This was originally proposed by @NightRa on Gitter.

>Played a little more with the plugin, it to parse output from stack on windows:
CabalConfigComponent.getResolverFromLocalPkgdb, stack 1.3.2 returns "C:\Users\Ilan\Programming\Haskell-Playground.stack-work\install\ef7d292a\pkgdb", which contains no version.
Also, assuming there must be an "lts-." is problematic - there are nightlies too

Currently this plugin support LTS and Nightly Stack resolver, but a hash value resolver may not relate to a specific cabal.config file.

### PR DESCRIPTION
**Pull request by zjhmale**
Fixes #125.

@rikvdkleij For a second thought, I believe there is no need to maintain a default cabal.config file in the plugin since each stack project should contain a validating resolver in its YAML file 😄

### CODE REVIEW

**Comment by rikvdkleij** [comment]
> @rikvdkleij For a second thought, I believe there is no need to maintain a default cabal.config file in the plugin since each stack project should contain a validating resolver in its YAML file :smile:

I think that is not correct, my stack.yaml contains"
```
# Resolver to choose a 'specific' stackage snapshot or a compiler version.
# A snapshot resolver dictates the compiler version and the set of packages
# to be used for project dependencies. For example:
# 
# resolver: lts-3.5
# resolver: nightly-2015-09-21
# resolver: ghc-7.10.2
# resolver: ghcjs-0.1.0_ghc-7.10.2
# resolver:
#  name: custom-snapshot
#  location: "./custom-snapshot.yaml"
resolver: lts-7.14

```

**Comment by zjhmale** [comment]
@rikvdkleij But the commented resolvers will not be retrieved, the on using resolver is lts-7.14.

**Comment by rikvdkleij** [comment]
@zjhmale But it can be for example `ghc-7.10.2`, and that one does not refer to cabal.config.

**Comment by zjhmale** [comment]
Ok, but everytime users reopen the project or restart stack repl, we will check whether the resolver from YAML file and cabal.config file are matching with each other.

**Comment by rikvdkleij** [comment]
But if there is no cabal.config downloaded because stack.yaml contains from beginning `ghc` resolver.

**Comment by zjhmale** [comment]
I got the point, sorry, I did not understand you at first 😅

**Comment by zjhmale** [comment]
@rikvdkleij I think in other resolvers the stack repl will also encounter some problems, e.g. with ghc-7.10.2 resolver.

image

image

Have to use `stack solver` and then add those suggested deps to stack.yaml to make everything works fine.

**Comment by zjhmale** [comment]
I think we should continue the original plan, after build intero manually, users can kindly use this plugin including package name completion with non lts or nightly resolvers. I just added notes on readme for reminding 😄 
>So no resolver but user still expects package name completion works 😄 Or we should not support that.

**Comment by zjhmale** [comment]
@rikvdkleij Please review, a default cabal.config is added 😄

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 1]
```diff
@@ -1,70 +1,88 @@
...
     downloadCabalConfig(project)
     parseCabalConfigFile(project)
   }
 
-  private def parseCabalConfigFile(project: Project): Option[Iterable[String]] = {
-    val configFilePath = getConfigFilePath(project)
+  private def parseCabalConfigFileBase(project: Project, lines: Iterator[String]): Option[Iterable[String]] = {
     try {
-      Some(Source.fromFile(configFilePath).getLines().filter(!_.startsWith("--")).map {
+      Some(lines.filter(!_.startsWith("--")).map {
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Why wrap it in Some?

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 2]
```diff
@@ -1,70 +1,88 @@
...
         case PackageNamePattern(packageName) => packageName
         case _ => ""
       }.filter(!_.isEmpty).toList)
     } catch {
       case _: FileNotFoundException =>
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Where does the FileNotFoundException come from?

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 3]
```diff
@@ -88,14 +105,18 @@ object CabalConfigComponent {
...
-        case _ => None
-      }
-    })
+  private def getResolverFromStackYamlFile(project: Project): Option[String] = {
+    try {
+      Option(new Yaml()
+        .load(new FileInputStream(new File(getYamlFilePath(project))))
+        .asInstanceOf[java.util.Map[String, Any]]
+        .get("resolver"))
+        .map(_.asInstanceOf[String])
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: What if value of "resolver" can not be found? get can return null.

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 4]
```diff
@@ -1,70 +1,88 @@
...
 
-  private def parseCabalConfigFile(project: Project): Option[Iterable[String]] = {
-    val configFilePath = getConfigFilePath(project)
+  private def parseCabalConfigFileBase(project: Project, lines: Iterator[String]): Option[Iterable[String]] = {
     try {
-      Some(Source.fromFile(configFilePath).getLines().filter(!_.startsWith("--")).map {
+      Some(lines.filter(!_.startsWith("--")).map {
         case PackageNamePattern(packageName) => packageName
         case _ => ""
       }.filter(!_.isEmpty).toList)
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Use filterNot

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 5]
```diff
@@ -1,70 +1,88 @@
...
+            } else {
+              parseCabalConfigFile(project)
+            }
+          })
         }
-      })
-    }
+      } else {
+        val a = parseDefaultCabalConfigFile(project)
+        a
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Refactor.

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala [hunk 6]
```diff
@@ -1,70 +1,88 @@
...
+        if (!configFile.exists()) {
           downloadAndParseCabalConfigFile(project)
         } else {
-          parseCabalConfigFile(project)
+          val needUpdate = for {
+            oldResolver <- getResolverFromCabalConfigFile(project)
+            newResolver <- getResolverFromStackYamlFile(project)
+          } yield oldResolver != newResolver
+
+          needUpdate.flatMap(b => {
```
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: What if needUpdate is None?

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala (hunk 5 above)
Comment (zjhmale) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Sorry, just for debug purpose, forgot to remove that 😅

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala (hunk 3 above)
Comment (zjhmale) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: Yes, get can return null, that's why I wrap it up with a `Option`, I will also parse default cabal config file if can not get `resolver` from Stack YAML file 😄

**Comment by zjhmale** [comment]
poke @rikvdkleij

File: src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala (hunk 3 above)
Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: But you are too late with wrapping in `Option` ....

Comment (rikvdkleij) [inline src/main/scala/intellij/haskell/external/component/CabalConfigComponent.scala]: O, wait, I do not saw second closing paren.

**Comment by rikvdkleij** [comment]
Thanks!

**Pull request merged**